### PR TITLE
refactoring

### PR DESCRIPTION
**Pull request by 0xIslamTaha**
Resolves: https://github.com/threefoldtech/jumpscaleX/issues/71

## Description
Refactor capacity.py and ubuntu.py files

### CODE REVIEW

File: Jumpscale/sal/ubuntu/Capacity.py [hunk 1]
```diff
@@ -56,19 +54,26 @@ def disk_info(self):
...
-        return j.tools.capacity.parser.get_report(psutil.virtual_memory().total, self.hw_info, self.disk_info, indent=indent)
+        return j.tools.capacity.parser.get_report(self._memory_info, self.hw_info, self.disk_info, indent=indent)
 
     def get(self, farmer_id):
         """
         get the capacity object of the node
 
         this capacity object is used in the capacity registration
 
+        :param: farmer_id
```
Comment (serboctor) [inline Jumpscale/sal/ubuntu/Capacity.py]: Docstring needs a brief description after the paramter, and it is missing the `type` like the convention

File: Jumpscale/sal/ubuntu/Capacity.py [hunk 2]
```diff
@@ -96,6 +101,13 @@ def get(self, farmer_id):
         return capacity
 
     def register(self, farmer_id):
+        """
+        register the node
+
+        :param farmer_id:
```
Comment (serboctor) [inline Jumpscale/sal/ubuntu/Capacity.py]: same as previous comment

File: Jumpscale/sal/ubuntu/Ubuntu.py [hunk 3]
```diff
@@ -43,9 +46,12 @@ def apt_init(self):
         self.aptCache = self._cache_ubuntu
```
Comment (serboctor) [inline Jumpscale/sal/ubuntu/Ubuntu.py]: change to `apt_cache`

File: Jumpscale/sal/ubuntu/Ubuntu.py [hunk 4]
```diff
@@ -63,86 +69,101 @@ def check(self, die=True):
...
-            rc, out, err = self._local.execute("which %s" % cmdname, False)
-            if rc != 0:
-                self.apt_install(packagename)
-            else:
-                return
-            rc, out, err = self._local.execute("which %s" % cmdname, False)
-            if rc != 0:
-                raise j.exceptions.RuntimeError(
-                    "Could not install package %s and check for command %s." % (packagename, cmdname))
+        rc, out, err = self._local.execute("which %s" % cmd_name, False)
```
Comment (serboctor) [inline Jumpscale/sal/ubuntu/Ubuntu.py]: the executors are going to be removed, so you need to replace them with python libraries like `subprocess` for example.
Same applies for the rest of the file.

**Review by serboctor — CHANGES_REQUESTED** [review]

**Comment by serboctor** [comment]
Also, please replace double quotes with single quotes while you're at it, except for function docstrings of course, so that the whole file is consistent.

File: Jumpscale/sal/ubuntu/Capacity.py [hunk 5]
```diff
@@ -56,19 +54,27 @@ def disk_info(self):
...
+        return j.tools.capacity.parser.get_report(self._memory_info, self.hw_info, self.disk_info, indent=indent)
 
     def get(self, farmer_id):
         """
         get the capacity object of the node
 
         this capacity object is used in the capacity registration
 
+        :param farmer_id: farmer id value
+        :rtype: str
```
Comment (serboctor) [inline Jumpscale/sal/ubuntu/Capacity.py]: this should be `:type farmer_id: str` not `:rtype: str`.....rtype is for the return type. Same applies for the rest of the file.

File: Jumpscale/sal/ubuntu/Capacity.py [hunk 6]
```diff
@@ -56,19 +54,27 @@ def disk_info(self):
                 )
         return self._disk_info
 
+    @property
+    def memory_info(self):
+        if self._memory_info is None:
+            self._memory_info = psutil.virtual_memory().total
+        return self._memory_info
+
```
Comment (serboctor) [inline Jumpscale/sal/ubuntu/Capacity.py:78]: some functions are still missing docstrings and lots of double quotes where single quotes are applicable.

File: Jumpscale/sal/ubuntu/Ubuntu.py [hunk 7]
```diff
@@ -35,228 +38,350 @@ def apt_init(self):
...
-            rc, out, err = self._local.execute("which %s" % cmdname, False)
-            if rc != 0:
-                self.apt_install(packagename)
-            else:
-                return
-            rc, out, err = self._local.execute("which %s" % cmdname, False)
-            if rc != 0:
-                raise j.exceptions.RuntimeError(
-                    "Could not install package %s and check for command %s." % (packagename, cmdname))
+        rc, out, err = self._local.execute('which %s' % cmd_name, False)
```
Comment (serboctor) [inline Jumpscale/sal/ubuntu/Ubuntu.py]: why did you remove the support for a list of packages?

Comment (serboctor) [inline Jumpscale/sal/ubuntu/Ubuntu.py]: please use sal.process...same comment for the rest of the file

File: Jumpscale/sal/ubuntu/Ubuntu.py [hunk 8]
```diff
@@ -35,228 +38,350 @@ def apt_init(self):
...
             j.tools.path.get(path).rmtree_p()
 
-    def pkg_list(self, pkgname, regex=""):
+    def pkg_list(self, pkg_name, regex=''):
         """
         list files of dpkg
         if regex used only output the ones who are matching regex
+
+        :param pkg_name: debian package name
+        :rtype: str
```
Comment (serboctor) [inline Jumpscale/sal/ubuntu/Ubuntu.py]: incorrect doc......same applies below

**Review by serboctor — CHANGES_REQUESTED** [review]
- There are no tests,
- Please use sal.process in the ubuntu file instead of the localexecutor
- There are still a lot of double quotes where single quotes can be used.
- Some functions dont have docstrings at all, and some docstrings are incorrect.

File: Jumpscale/sal/ubuntu/tests/test_ubuntu.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    def test014_sshkey_generate(self):
+        self.ubuntu.sshkey_generate(path='/tmp/id_rsa')
+        rc, out, err = j.sal.process.execute('ls /tmp | grep id_rsa')
+        self.assertIn('id_rsa', out)
+
+
+def test_main(self=None):
+    """
+    to run:
+    js_shell 'j.sal.tls._test(name="tls")'
```
Comment (serboctor) [inline Jumpscale/sal/ubuntu/tests/test_ubuntu.py]: wrong description

**Review by serboctor — CHANGES_REQUESTED** [review]

File: Jumpscale/sal/ubuntu/Ubuntu.py [hunk 10]
```diff
@@ -35,239 +40,303 @@ def apt_init(self):
...
-        result = []
-        for item in self._cache_ubuntu.keys():
-            item2 = item.replace("_", "").replace("_", "").lower()
-            if item2.find(packagename) != -1:
-                result.append(item)
-        return result
-
-    def get_installed_package_names(self):
+        :return: list of installed list
+        :type: list
```
Comment (serboctor) [inline Jumpscale/sal/ubuntu/Ubuntu.py]: this should be rtype

File: Jumpscale/sal/ubuntu/Ubuntu.py [hunk 11]
```diff
@@ -278,66 +347,86 @@ def get_installed_package_names(self):
...
         return self._installed_pkgs
 
-    def is_pkg_installed(self, pkg):
-        return pkg in self._installed_pkgs
+    def apt_find_all(self, package_name):
+        """find all packages match with the package_name
 
-    def apt_find_installed(self, packagename):
-        packagename = packagename.lower().strip().replace("_", "").replace("_", "")
+        :param package_name: ubuntu package name
```
Comment (serboctor) [inline Jumpscale/sal/ubuntu/Ubuntu.py:352]: missing type for package name

File: Jumpscale/sal/ubuntu/Ubuntu.py [hunk 12]
```diff
@@ -278,66 +347,86 @@ def get_installed_package_names(self):
...
-            return res[0]
-        elif len(res) > 1:
-            raise j.exceptions.RuntimeError("Found more than 1 package for %s" % packagename)
-        raise j.exceptions.RuntimeError("Could not find package %s" % packagename)
+    def is_pkg_installed(self, package_name):
+        """check if the package is installed or not.
+
+        :param package_name: package name
+        :type: str
+        :return: bool
```
Comment (serboctor) [inline Jumpscale/sal/ubuntu/Ubuntu.py]: this should be a description, and there should be `rtype: bool`

File: Jumpscale/sal/ubuntu/Capacity.py [hunk 13]
```diff
@@ -56,19 +64,33 @@ def disk_info(self):
...
     def report(self, indent=None):
         """
         create a report of the hardware capacity for
         processor, memory, motherboard and disks
-        """
-        return j.tools.capacity.parser.get_report(psutil.virtual_memory().total, self.hw_info, self.disk_info, indent=indent)
 
-    def get(self, farmer_id):
+        :return: report for the hardware capacity information
```
Comment (serboctor) [inline Jumpscale/sal/ubuntu/Capacity.py:86]: docstring is missing the `indent parameter`

File: Jumpscale/sal/ubuntu/Capacity.py (hunk 6 above)
Comment (0xIslamTaha) [reply Jumpscale/sal/ubuntu/Capacity.py:78]: fixed

File: Jumpscale/sal/ubuntu/Capacity.py (hunk 13 above)
Comment (0xIslamTaha) [reply Jumpscale/sal/ubuntu/Capacity.py:86]: done

File: Jumpscale/sal/ubuntu/Ubuntu.py (hunk 10 above)
Comment (0xIslamTaha) [reply Jumpscale/sal/ubuntu/Ubuntu.py]: done

File: Jumpscale/sal/ubuntu/Ubuntu.py (hunk 11 above)
Comment (0xIslamTaha) [reply Jumpscale/sal/ubuntu/Ubuntu.py:352]: done

File: Jumpscale/sal/ubuntu/Ubuntu.py (hunk 12 above)
Comment (0xIslamTaha) [reply Jumpscale/sal/ubuntu/Ubuntu.py]: done
